src/puppetlabs/puppetdb/catalogs.clj

```diff
@@ -367,13 +367,32 @@
            (throw (IllegalArgumentException. (format "Catalog must be specified as a string or a map, not '%s'" (class catalog)))))))
 
 (defmethod parse-catalog String
-  [catalog version]
+  [catalog version received-time]
   {:pre   [(string? catalog)]
    :post  [(map? %)]}
-  (parse-catalog (json/parse-string catalog true) version))
+  (parse-catalog (json/parse-string catalog true) version received-time))
+
+(defmethod parse-catalog 4
+  [catalog version received-time]
+  {:pre [(map? catalog)
+         (number? version)]
+   :post [(map? %)]}
+  (-> catalog
+      (assoc :producer_timestamp received-time)
```

Comment:
I'm a bit concerned that post-upgrade we're substituting (now) for the proper producer_timestamp, and that this will cause the newer catalogs (with timestamps set by the master, before the upgrade occurred) to be rejected in favor of the old ones